File: solutions/fable-1/fable/typescript/dijkstra.ts
const data = require("fs").readFileSync(0, "utf8");
const tokens = data.split(/\s+/).filter((x: string) => x.length > 0).map(Number);
let p = 0;
const n = tokens[p++];
const m = tokens[p++];

// CSR adjacency
const deg = new Int32Array(n + 1);
const eu = new Int32Array(m);
const ev = new Int32Array(m);
const ew = new Float64Array(m);
for (let i = 0; i < m; i++) {
  const u = tokens[p++];
  const v = tokens[p++];
  const w = tokens[p++];
  eu[i] = u;
  ev[i] = v;
  ew[i] = w;
  deg[u + 1]++;
  deg[v + 1]++;
}
for (let i = 0; i < n; i++) deg[i + 1] += deg[i];
const pos = deg.slice(0, n);
const adjTo = new Int32Array(2 * m);
const adjW = new Float64Array(2 * m);
for (let i = 0; i < m; i++) {
  adjTo[pos[eu[i]]] = ev[i];
  adjW[pos[eu[i]]++] = ew[i];
  adjTo[pos[ev[i]]] = eu[i];
  adjW[pos[ev[i]]++] = ew[i];
}

const s = tokens[p++];
const t = tokens[p++];

// Binary heap of (dist, node)
let heapD = new Float64Array(1024);
let heapN = new Int32Array(1024);
let hsize = 0;
function push(d: number, v: number) {
  if (hsize === heapD.length) {
    const nd = new Float64Array(hsize * 2);
    nd.set(heapD);
    heapD = nd;
    const nn = new Int32Array(hsize * 2);
    nn.set(heapN);
    heapN = nn;
  }
  let i = hsize++;
  while (i > 0) {
    const par = (i - 1) >> 1;
    if (heapD[par] <= d) break;
    heapD[i] = heapD[par];
    heapN[i] = heapN[par];
    i = par;
  }
  heapD[i] = d;
  heapN[i] = v;
}
function pop(): void {
  hsize--;
  if (hsize === 0) return;
  const d = heapD[hsize];
  const v = heapN[hsize];
  let i = 0;
  while (true) {
    let c = 2 * i + 1;
    if (c >= hsize) break;
    if (c + 1 < hsize && heapD[c + 1] < heapD[c]) c++;
    if (heapD[c] >= d) break;
    heapD[i] = heapD[c];
    heapN[i] = heapN[c];
    i = c;
  }
  heapD[i] = d;
  heapN[i] = v;
}

const dist = new Float64Array(n).fill(Infinity);
let answer = -1;
if (s === t) {
  answer = 0;
} else if (s >= 0 && s < n && t >= 0 && t < n) {
  dist[s] = 0;
  push(0, s);
  while (hsize > 0) {
    const d = heapD[0];
    const u = heapN[0];
    pop();
    if (d > dist[u]) continue;
    if (u === t) break;
    for (let k = deg[u]; k < deg[u + 1]; k++) {
      const v = adjTo[k];
      const nd = d + adjW[k];
      if (nd < dist[v]) {
        dist[v] = nd;
        push(nd, v);
      }
    }
  }
  if (dist[t] !== Infinity) answer = dist[t];
}
process.stdout.write(String(answer) + "\n");

File: solutions/fable-1/fable/zig/dijkstra.zig
const std = @import("std");

const Item = struct {
    dist: u64,
    node: u32,
};

fn lessThan(_: void, a: Item, b: Item) std.math.Order {
    return std.math.order(a.dist, b.dist);
}

const Parser = struct {
    buf: []const u8,
    pos: usize = 0,

    fn next(self: *Parser) ?u64 {
        while (self.pos < self.buf.len and (self.buf[self.pos] < '0' or self.buf[self.pos] > '9')) : (self.pos += 1) {}
        if (self.pos >= self.buf.len) return null;
        var v: u64 = 0;
        while (self.pos < self.buf.len and self.buf[self.pos] >= '0' and self.buf[self.pos] <= '9') : (self.pos += 1) {
            v = v * 10 + (self.buf[self.pos] - '0');
        }
        return v;
    }
};

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const stdin = std.fs.File.stdin();
    const input = try stdin.readToEndAlloc(alloc, 1 << 30);

    var p = Parser{ .buf = input };
    const n: usize = @intCast(p.next() orelse 0);
    const m: usize = @intCast(p.next() orelse 0);

    const eu = try alloc.alloc(u32, m);
    const ev = try alloc.alloc(u32, m);
    const ew = try alloc.alloc(u64, m);
    const deg = try alloc.alloc(u32, n + 1);
    @memset(deg, 0);

    var i: usize = 0;
    while (i < m) : (i += 1) {
        const u: u32 = @intCast(p.next() orelse 0);
        const v: u32 = @intCast(p.next() orelse 0);
        const w: u64 = p.next() orelse 0;
        eu[i] = u;
        ev[i] = v;
        ew[i] = w;
        if (u < n) deg[u] += 1;
        if (v < n) deg[v] += 1;
    }
    const s: usize = @intCast(p.next() orelse 0);
    const t: usize = @intCast(p.next() orelse 0);

    // CSR
    const off = try alloc.alloc(u32, n + 1);
    off[0] = 0;
    i = 0;
    while (i < n) : (i += 1) off[i + 1] = off[i] + deg[i];
    const total = off[n];
    const adjTo = try alloc.alloc(u32, total);
    const adjW = try alloc.alloc(u64, total);
    const fill = try alloc.alloc(u32, n);
    @memcpy(fill, off[0..n]);
    i = 0;
    while (i < m) : (i += 1) {
        const u = eu[i];
        const v = ev[i];
        if (u >= n or v >= n) continue;
        adjTo[fill[u]] = v;
        adjW[fill[u]] = ew[i];
        fill[u] += 1;
        adjTo[fill[v]] = u;
        adjW[fill[v]] = ew[i];
        fill[v] += 1;
    }

    const stdout = std.fs.File.stdout();

    if (s == t) {
        try stdout.writeAll("0\n");
        return;
    }
    if (s >= n or t >= n) {
        try stdout.writeAll("-1\n");
        return;
    }

    const inf = std.math.maxInt(u64);
    const dist = try alloc.alloc(u64, n);
    @memset(dist, inf);
    dist[s] = 0;

    var heap = std.PriorityQueue(Item, void, lessThan).init(alloc, {});
    try heap.add(.{ .dist = 0, .node = @intCast(s) });

    while (heap.removeOrNull()) |it| {
        if (it.dist > dist[it.node]) continue;
        if (it.node == t) break;
        const u = it.node;
        var k = off[u];
        const end = off[u + 1];
        while (k < end) : (k += 1) {
            const v = adjTo[k];
            const nd = it.dist + adjW[k];
            if (nd < dist[v]) {
                dist[v] = nd;
                try heap.add(.{ .dist = nd, .node = v });
            }
        }
    }

    var outbuf: [64]u8 = undefined;
    if (dist[t] == inf) {
        try stdout.writeAll("-1\n");
    } else {
        const str = try std.fmt.bufPrint(&outbuf, "{d}\n", .{dist[t]});
        try stdout.writeAll(str);
    }
}
